eth/tracers/firehose.go

```diff
@@ -280,12 +291,46 @@ func NewFirehose(config *FirehoseConfig) *Firehose {
 		}
 	}
 
-	if config.ConcurrentBlockFlushing {
-		log.Info("Firehose concurrent block flushing enabled, starting block " +
-			"print worker goroutine")
-		firehose.blockPrintQueue = make(chan *blockPrintJob, 100)
-		firehose.blockFlushDone.Add(1)
-		go firehose.blockPrintWorker()
+	if config.ConcurrentBlockFlushing > 0 {
+		log.Info(fmt.Sprintf("Firehose concurrent block flushing enabled, starting %d worker goroutine", config.ConcurrentBlockFlushing))
+
+		firehose.flushJobQueue = make(chan *blockPrintJob, firehose.flushBufferSize)
+		firehose.flushOrderedOutputQueue = make(chan *blockOutput, firehose.flushBufferSize)
+		for i := 0; i < config.ConcurrentBlockFlushing; i++ {
+			firehose.flushJobWG.Add(1)
+			go firehose.blockPrintWorker()
+		}
+
+		// Output channel to order the flushing linearly
+		firehose.flushOrderedOutputWG.Add(1)
+		go func() {
+			defer firehose.flushOrderedOutputWG.Done()
+
+			buffer := make(map[uint64][]byte)
+
+			firehose.flushStartCond.L.Lock()
+			for !firehose.flushStarted {
+				firehose.flushStartCond.Wait()
+			}
+
+			nextExpected := firehose.flushStartBlockNum
+			firehose.flushStartCond.L.Unlock()
+
+			for result := range firehose.flushOrderedOutputQueue {
+				buffer[result.blockNum] = result.data
+
+				for {
+					data, ok := buffer[nextExpected]
+					if !ok {
+						break
+					}
+
+					firehose.flushToFirehose(data)
+					delete(buffer, nextExpected)
+					nextExpected++
+				}
+			}
+		}()
 	}
```

Comment:
This seems wrong to be in the constructor, I feel it should go in `BlockchainInit` instead, which is where the tracer is actually started.

Also, the previous comment about a `FlushQueue` make total sense. Maybe we should explorer existing package actually. https://github.com/destel/rill is a library we looked at.